api/py/ai/chronon/utils.py

```diff
@@ -384,67 +402,86 @@ def requires_log_flattening_task(conf: ChrononJobTypes) -> bool:
-    modes = []  # type: List[str]
+def get_modes_tables(conf: ChrononJobTypes) -> List[str]:
+    """Based on a conf get all the applicable modes and tables"""
+
+    if conf.metaData.name:
+        table_name = output_table_name(conf, full_name=True)
+    else:
+        table_name = output_table_name(conf, full_name=False)
+
+    modes = []
+    tables = {}
 
+    # from GroupBy
     if isinstance(conf, api.GroupBy):
         group_by = cast(api.GroupBy, conf)
         if group_by.backfillStartDate is not None:
-            modes.append("backfill")
+            modes.append(Modes.backfill)
+            tables[Modes.backfill] = [f"{table_name}"]
 
         online = group_by.metaData.online or False
-
         if online:
-            modes.append("upload")
+            modes.append(Modes.upload)
+            tables[Modes.upload] = [f"{table_name}_upload"]
 
             temporal_accuracy = group_by.accuracy or False
             streaming = has_topic(group_by)
             if temporal_accuracy or streaming:
-                modes.append("streaming")
+                modes.append(Modes.streaming)
+    # from Join
     elif isinstance(conf, api.Join):
         join = cast(api.Join, conf)
         if get_offline_schedule(conf) is not None:
-            modes.append("backfill")
-            modes.append("stats-summary")
+            modes.append(Modes.backfill)
+            modes.append(Modes.stats_summary)
+            tables[Modes.backfill] = [f"{table_name}"]
+            tables[Modes.stats_summary] = [f"{table_name}_daily_stats"]
         if (
             join.metaData.customJson is not None
             and json.loads(join.metaData.customJson).get("check_consistency") is True
         ):
-            modes.append("consistency-metrics-compute")
+            modes.append(Modes.consistency_metrics_compute)
+            tables[Modes.consistency_metrics_compute] = [f"{table_name}_consistency"]
         if requires_log_flattening_task(join):
-            modes.append("log-flattener")
+            modes.append(Modes.log_flattener)
+            tables[Modes.log_flattener] = [f"{table_name}_logged"]
         if join.labelPart is not None:
-            modes.append("label-join")
+            modes.append(Modes.label_join)
+            tables[Modes.label_join] = [f"{table_name}_labels", f"{table_name}_labeled", f"{table_name}_labeled_latest"]
+
+        if conf.bootstrapParts:
+            tables[SubStage.bootstrap] = [f"{table_name}_bootstrap"]
+
+        if conf.joinParts:
+            tables[SubStage.join_parts] = [join_part_output_table_name(conf, jp) for jp in conf.joinParts]
+    # from StagingQuery
     elif isinstance(conf, api.StagingQuery):
-        modes.append("backfill")
+        modes.append(Modes.backfill)
+        tables[Modes.backfill] = [f"{table_name}"]
     else:
         raise ValueError(f"Unsupported job type {type(conf).__name__}")
 
+    return modes, tables
```

Comment:
We are collecting data in 2 separate variables because of a filtering condition here (i.e. not include Substage). My suggestion is to make that filtering explicit in the caller (get_applicable_modes).
Return only tables from get_modes_tables and caller can filter from dict.keys().

get_related_table_names seems to filter SubStage anyway?

Reply:
ah I see. Makes sense.

Reply:
The problem here is that there are some modes which do not produce tables. (e.g. Streaming).

But I think it's better to have a 'mode': None when this occurs over 2 separate outputs.

Reply:
changed